index.bs

```diff
@@ -1322,6 +1548,111 @@ To <dfn for="MLGraphBuilder">validate operand</dfn> given {{MLGraphBuilder}} |bu
+  readonly attribute boolean writable;
+
+  undefined destroy();
+};
+</script>
+
+<div class=internal-slots>
+{{MLTensor}} has the following internal slots:
+  <dl dfn-type=attribute dfn-for="MLTensor">
+    : <dfn>\[[context]]</dfn> of type {{MLContext}}
+    ::
+        The {{MLTensor}}'s associated context.
+
+    : <dfn>\[[descriptor]]</dfn> of type {{MLTensorDescriptor}}
+    ::
+        The {{MLTensor}}'s descriptor.
+
+    : <dfn>\[[isDestroyed]]</dfn> of type {{boolean}}
+    ::
+        Whether {{MLTensor}}.{{MLTensor/destroy()}} has been called. Once destroyed, the {{MLTensor}} can no longer be used.
+
+    : <dfn>\[[data]]</dfn> of an [=implementation-defined=] type
+    ::
+        The bytes backing the {{MLTensor}}. This data may only be accessed or modified from the {{MLTensor/[[context]]}}'s {{MLContext/[[timeline]]}}.
+  </dl>
+</div>
+
+An {{MLTensor}}'s <dfn for=MLTensor>dataType</dfn> is its {{MLTensor/[[descriptor]]}}'s {{MLOperandDescriptor/dataType}}.
+
+An {{MLTensor}}'s <dfn for=MLTensor>shape</dfn> is its {{MLTensor/[[descriptor]]}}'s {{MLOperandDescriptor/shape}}.
+
+The <dfn attribute for=MLTensor>dataType</dfn> [=getter steps=] are to return [=this=]'s [=MLTensor/dataType=].
+
+The <dfn attribute for=MLTensor>shape</dfn> [=getter steps=] are to return [=this=]'s [=MLTensor/shape=].
+
+The <dfn attribute for=MLTensor>readable</dfn> [=getter steps=] are to return [=this=].{{MLTensor/[[descriptor]]}}["{{MLTensorDescriptor/readable}}"].
+
+The <dfn attribute for=MLTensor>writable</dfn> [=getter steps=] are to return [=this=].{{MLTensor/[[descriptor]]}}["{{MLTensorDescriptor/writable}}"].
+
+### Creating an {{MLTensor}} ### {#api-mltensor-create}
+
+An {{MLTensor}} is created by its associated {{MLContext}}.
+
+<details open algorithm>
+  <summary>
+    To <dfn>create an MLTensor</dfn> given {{MLContext}} |context| and {{MLTensorDescriptor}} |descriptor|, run the following steps:
+  </summary>
+    1. Let |tensor| be a new {{MLTensor}}.
+    1. Set |tensor|.{{MLTensor/[[context]]}} to |context|.
+    1. Set |tensor|.{{MLTensor/[[descriptor]]}} to |descriptor|.
+    1. Set |tensor|.{{MLTensor/[[isDestroyed]]}} to false.
+    1. Return |tensor|.
+</details>
+
+### {{MLTensor/destroy()}} ### {#api-mltensor-destroy}
+
+Frees the resources associated with the {{MLTensor}}. This method is idempotent.
+
+<div dfn-for="MLTensor/destroy()" dfn-type=argument>
+    **Returns:** {{undefined}}.
+</div>
+
+<details open algorithm>
+  <summary>
+    The <dfn method for=MLTensor>destroy()</dfn> method steps are:
+  </summary>
+    1. Set [=this=].{{MLTensor/[[isDestroyed]]}} to true.
+    1. Enqueue the following steps to [=this=].{{MLTensor/[[context]]}}'s {{MLContext/[[timeline]]}}:
+        1. Free [=this=].{{MLTensor/[[data]]}}.
+    1. Return undefined.
+</details>
+
+Note: Since no further operations can be enqueued using this tensor, implementations can free any additional resource allocations associated with this tensor.
```

Comment:
I don't think that's necessary, especially since resources will be freed on _both the context and content timelines_. Please feel free to unresolve if you disagree

Reply:
It seems not clear to me what the additional resources are. If they are allocated on GPU/NPU device, for example, the staging buffers for uploading/reading-back, should they be released on context's timeline?

Reply:
Right, but the resources used to implement an `MLTensor` (such as staging buffers) are an implementation detail, and once the tensor can no longer be used, it shouldn't matter what timeline the resources are released on (or if they're released at all) since it's not observable

This is a non-normative note anyways so I don't think we need to be dictating timelines here?

Reply:
> once the tensor can no longer be used, it shouldn't matter what timeline the resources are released on (or if they're released at all) since it's not observable

The already enqueued tensor write/read operations may still need to access the staging buffers.

Reply:
I'm hesitant to get too in the weeds in a non-normative note. From the perspective of the spec, what's important is that this `MLTensor` can no longer be used, so UAs are can release its associated resources as they see fit. Of course some of those resources might be bound up in ongoing operations, but from the perspective of the spec that's an implementation detail we trust the UA to figure out without running into UAFs.

FWIW this text was lifted directly from [the `GPUBuffer.destroy()` method of the WebGPU spec](https://www.w3.org/TR/webgpu/#dom-gpubuffer-destroy), which this method is modeled after 😛

Reply:
> I feel "release" may be a better term

Agreed, updated the algorithm

> BTW, I found [GPUBuffer Destruction](https://www.w3.org/TR/webgpu/#buffer-destruction) has a note for GPU resource:
> 
> > Note: This allows the user agent to reclaim the GPU memory associated with the GPUBuffer once all previously submitted operations using it are complete.

Ah yes, good call out. I've amended the note here to reflect that... WDYT?

Reply:
Looks good! Thanks for your patience.